AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStoreAdmin/src/CreateKeysHV2.dfy

```diff
@@ -58,166 +71,37 @@ module {:options "/functionSyntax:4" } CreateKeysHV2 {
-              && writeNewEncryptedBranchKey.input.Active
-                 == Structure.ConstructEncryptedHierarchicalKey(
-                      Structure.ActiveBranchKeyEncryptionContext(decryptOnlyBranchKeyContext),
-                      kmsEncryptRequestForActiveBKI.output.value.CiphertextBlob.value
-                    )
-              && writeNewEncryptedBranchKey.input.Version
-                 == Structure.ConstructEncryptedHierarchicalKey(
-                      decryptOnlyBranchKeyContext,
-                      kmsEncryptRequestForDecryptOnlyBKI.output.value.CiphertextBlob.value
-                    )
-              && writeNewEncryptedBranchKey.input.Beacon
-                 == Structure.ConstructEncryptedHierarchicalKey(
-                      Structure.BeaconKeyEncryptionContext(decryptOnlyBranchKeyContext),
-                      kmsEncryptRequestForBeaconBKI.output.value.CiphertextBlob.value
-                    )
-
-              && writeNewEncryptedBranchKey.output.Success?
-              && output.value.Identifier == branchKeyIdentifier
-    //= aws-encryption-sdk-specification/framework/branch-key-store.md#createkey
-    //= type=implication
-    //# Otherwise, this operation MUST yield an error.
+    // Generic Stateful pre and post conditions
+    requires
+      && kmsClient.ValidState()
+      && storage.ValidState()
+      && kmsClient.Modifies !! storage.Modifies
+    modifies multiset(storage.Modifies) + multiset(kmsClient.Modifies)
     ensures
-      // TODO-HV-2-FOLLOW : Is there another way to represent this? This is also not nearly all the possible failrues...
-      || (&& |kmsClient.History.Encrypt| == |old(kmsClient.History.Encrypt)| + 1
-          && Seq.Last(kmsClient.History.Encrypt).output.Failure?
-          ==> output.Failure?)
-
-      || (&& |kmsClient.History.Encrypt| == |old(kmsClient.History.Encrypt)| + 2
-          && Seq.Last(Seq.DropLast(kmsClient.History.Encrypt)).output.Failure?
-          ==> output.Failure?)
-
-      || (&& |kmsClient.History.Encrypt| == |old(kmsClient.History.Encrypt)| + 3
-          && Seq.Last(Seq.DropLast(Seq.DropLast(kmsClient.History.Encrypt))).output.Failure?
-          ==> output.Failure?)
-
-      || (&& |storage.History.WriteNewEncryptedBranchKey| == |old(storage.History.WriteNewEncryptedBranchKey)| + 1
-          && Seq.Last(storage.History.WriteNewEncryptedBranchKey).output.Failure?
-          ==> output.Failure?)
-
+      && storage.ValidState()
+      && kmsClient.ValidState()
+      && kmsClient.Modifies !! storage.Modifies
+    // 
+    requires
+      && KMSKeystoreOperations.HasKeyId(kmsConfiguration)
+      && KmsArn.ValidKmsArn?(KMSKeystoreOperations.GetKeyId(kmsConfiguration))
+      && hierarchyVersion.v2?
+      && 0 < |branchKeyIdentifier|
+      && 0 < |branchKeyVersion|
+      && forall k <- encryptionContext :: DDB.Types.IsValid_AttributeName(Structure.ENCRYPTION_CONTEXT_PREFIX + k)
```

Comment:
Stylistically, should this be put in front of the preceding ensures clause?

Reply:
yes it should

Reply:
I thought we kept the stateful specifications together?
But I can create a datatype that will simplify the storage and KMS Client relationship,
and then resolve this.